### PR TITLE
Church-encoded State carrier

### PR DESCRIPTION
**Pull request by robrix**
This PR defines a Church-encoded `State` carrier, equivalent to `Codensity ((->) s)`.

I mostly wanted to see how the `MonadFix` instance would turn out, since I’m sort of combining this and the `CutC` instance for `fused-effects-parser`.

- [x] Depends on #366.
- [x] Update for the new shape of `Algebra`, GADT effects, &c.
- [x] Link from `Control.Effect.State`.

### CODE REVIEW

File: src/Control/Carrier/State/Church.hs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+instance MonadTrans (StateC s) where
+  lift m = StateC $ \ k s -> m >>= flip k s
+  {-# INLINE lift #-}
+
+instance (Algebra sig m, Effect sig) => Algebra (State s :+: sig) (StateC s m) where
+  alg hom = \case
+    L (Get   k) -> StateC $ \ k' s -> runState s (hom (k s)) >>= uncurry (flip k')
+    L (Put s k) -> StateC $ \ k' _ -> runState s (hom k)     >>= uncurry (flip k')
+    R other     -> StateC $ \ k  s -> alg id (thread (s, ()) (uncurry runState . fmap hom) other) >>= uncurry (flip k)
```
Comment (robrix) [inline src/Control/Carrier/State/Church.hs]: This turned out pretty nicely, but I’m not completely convinced that this is the best way to be threading the state through. I plan to revisit that in/after #361.

**Comment by patrickt** [comment]
I think this is a fine addition, though I think we should document something about its performance. Do we expect it to be better or worse than the `s -> (s, a)` form?

**Comment by robrix** [comment]
@patrickt: I don’t have a strong intuition, and in fact I don’t even think there’s going to be a cut and dried answer to that. My guess is that it’d depend pretty crucially on the carriers surrounding it. My goal is to try to provide some more church-encoded carriers and measure how they fare against the classical varieties in larger carrier compositions in e.g. Starlight or `semantic`.

**Comment by patrickt** [comment]
@robrix That’s fine by me. We should probably mention somewhere that the choice of the “best” carrier for state is contextual.

**Comment by robrix** [comment]
Yeah, I might go a bit further and recommend `Strict.StateC` as a default for the time being until we know more.

File: src/Control/Carrier/State/Church.hs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+-- 'execState' s m = 'fmap' 'fst' ('runState' s m)
+-- @
+--
+-- @since 1.1.0.0
+execState :: forall s m a . Applicative m => s -> StateC s m a -> m s
+execState s (StateC m) = m (const pure) s
+{-# INLINE execState #-}
+
+-- | @since 1.1.0.0
+newtype StateC s m a = StateC { runStateC :: forall r . (a -> s -> m r) -> s -> m r }
```
[the remark this answers is not in the document]
Comment (patrickt) [reply src/Control/Carrier/State/Church.hs]: I wouldn’t be averse to exporting the constructor, fwiw.

[the remark this answers is not in the document]
Comment (robrix) [reply src/Control/Carrier/State/Church.hs]: I do export the constructor, just not the selector.

**Pull request merged**